revolution_pi/gpio_pin.go

```diff
@@ -76,26 +132,52 @@ func (pin *gpioPin) Set(ctx context.Context, high bool, extra map[string]interfa
 
 // Get gets the high/low state of the pin.
 func (pin *gpioPin) Get(ctx context.Context, extra map[string]interface{}) (bool, error) {
-	pin.ControlChip.logger.Debugf("Reading from %v", pin.getGpioAddress())
-
 	if !pin.initialized {
 		return false, errors.New("pin not initialized")
 	}
-	return pin.ControlChip.getBitValue(int64(pin.getGpioAddress()), pin.BitPosition)
+	if pin.pwmMode {
+		return false, fmt.Errorf("cannot get pin state, Pin %s is configured as PWM", pin.Name)
+	}
+
+	gpioAddress := pin.getGpioAddress()
+	gpioBit := pin.BitPosition
+
+	if pin.isOutputPWM() {
+		// if someone used the PWM pin name, get the get the bit for the pin
+		gpioBit = uint8((pin.Address-outputWordToPWMOffset)%pin.outputOffset) % 8
+	} else if pin.isInputCounter() {
+		// if someone used the Counter pin name, get the get the bit for the pin
+		// get the address into 4 byte chunks, then mod 8 for the bit location
+		gpioBit = uint8((pin.Address-inputWordToCounterOffset)%pin.inputOffset) / 4 % 8
+	}
+
+	pin.ControlChip.logger.Debugf("Reading from %d", gpioAddress)
+	pin.ControlChip.logger.Debugf("Reading from bit %d", gpioBit)
+
+	return pin.ControlChip.getBitValue(int64(gpioAddress), gpioBit)
 }
 
 // PWM gets the pin's given duty cycle.
 func (pin *gpioPin) PWM(ctx context.Context, extra map[string]interface{}) (float64, error) {
-	pwmAddress := pin.getPwmAddress()
-	pin.ControlChip.logger.Debugf("Reading from %v", pwmAddress)
-
 	if !pin.initialized {
 		return 0, errors.New("pin not initialized")
 	}
+	if !pin.isOutputPWM() && !pin.isOutputWord() {
+		return 0, fmt.Errorf("cannot get PWM, Pin %s is not a PWM pin", pin.Name)
```

Comment:
I still don't really understand what `isOutputWord` means, but I wonder if that's a necessary part of the check. If this pin is not an output PWM, isn't that enough to say we can't get PWM stuff? 

Same comment on line 200

Reply:
This is more aliasing stuff. On line 200 I wanted to make the distinction between PWMs being invalid and PWMs not being configured